dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java

```diff
@@ -60,16 +64,28 @@ public List<BlockWorkerInfo> getPreferredWorkers(WorkerClusterView workerCluster
           "Not enough workers in the cluster %d workers in the cluster but %d required",
           workerClusterView.size(), count));
     }
-    List<BlockWorkerInfo> blockWorkerInfoList = workerClusterView.stream()
-        .map(w -> new BlockWorkerInfo(w.getIdentity(), w.getAddress(), w.getCapacityBytes(),
-            w.getUsedBytes(), w.getState() == WorkerState.LIVE))
+    List<WorkerIdentity> workerIdentities = workerClusterView.stream()
+        .map(WorkerInfo::getIdentity)
         .collect(Collectors.toList());
-    HASH_PROVIDER.refresh(blockWorkerInfoList, mNumVirtualNodes);
-    List<BlockWorkerInfo> workers = HASH_PROVIDER.getMultiple(fileId, count);
+    mHashProvider.refresh(workerIdentities, mNumVirtualNodes);
+    List<WorkerIdentity> workers = mHashProvider.getMultiple(fileId, count);
     if (workers.size() != count) {
       throw new ResourceExhaustedException(String.format(
           "Found %d workers from the hash ring but %d required", workers.size(), count));
     }
-    return workers;
+    ImmutableList.Builder<BlockWorkerInfo> builder = ImmutableList.builder();
+    for (WorkerIdentity worker : workers) {
+      WorkerInfo workerInfo = workerClusterView.getWorkerById(worker)
+          .orElseThrow(() -> new IllegalStateException(
+              String.format("Hash provider returned a non-existent worker: %s, "
+                  + "workers currently known: %s", worker, workerIdentities)));
```

Comment:
This can happen since `ConsistentHashProvider` has an internal cache of the list of workers, which has a TTL of 1 s. if two concurrent calls to `mHashProvider.refresh` with different lists of workers happen within the TTL, only one will actually update the hash ring, and the other will use either a stale view of workers, or use a view from the future and cause inconsistencies with the worker info list the caller currently has.